lib/unison-sqlite/src/Unison/Sqlite/Exception.hs

```diff
@@ -1,75 +1,133 @@
--- | Sqlite exception utils
+-- | Sqlite exception utils.
 module Unison.Sqlite.Exception
-  ( SqliteException (..),
-    SqliteExceptionReason,
+  ( -- * @SomeSqliteException@
+    SomeSqliteException (..),
+
+    -- ** @SqliteConnectException@
+    SqliteConnectException (..),
+    rethrowAsSqliteConnectException,
+
+    -- ** @SqliteQueryException@
+    SqliteQueryException (..),
+    SqliteQueryExceptionInfo (..),
+    throwSqliteQueryException,
     SomeSqliteExceptionReason (..),
-    SqliteExceptionInfo (..),
-    throwSqliteException,
+    SqliteExceptionReason,
   )
 where
 
 import Control.Concurrent (ThreadId, myThreadId)
+import Data.Typeable (cast)
 import qualified Database.SQLite.Simple as Sqlite
 import Debug.RecoverRTTI (anythingToString)
 import Unison.Prelude
 import Unison.Sqlite.Sql
 import UnliftIO.Exception
 
-data SqliteExceptionInfo params connection = SqliteExceptionInfo
-  { sql :: Sql,
+------------------------------------------------------------------------------------------------------------------------
+-- SomeSqliteException
+
+-- | The root exception for all exceptions thrown by this library.
+--
+-- @
+-- SomeException (from base)
+--   └── SomeSqliteException
+--         └── SqliteConnectException
+--         └── SqliteQueryException
+-- @
+--
+-- A @SomeSqliteException@ should not be inspected or used for control flow when run in a trusted environment, where the
+-- database can be assumed to be uncorrupt. Rather, wherever possible, the user of this library should write code that
+-- is guaranteed not to throw exceptions, by checking the necessary preconditions first. If that is not possible, it
+-- should be considered a bug in this library.
+--
+-- When actions are run on an untrusted codebase, e.g. one downloaded from a remote server, it is sufficient to catch
+-- just one exception type, @SomeSqliteException@.
+data SomeSqliteException
+  = forall e. Exception e => SomeSqliteException e
+  deriving anyclass (Exception)
```

Comment:
I believe we want a custom `fromException`/`toException` for `SomeSqliteException` don't we?

Have you tested to ensure that everything we expect is properly caught by this?

Looks like the default `fromException` just uses `cast` which I don't think would work properly for converting things like `SqliteConnectException`

Reply:
`SomeSqliteException` is directly beneath `SomeException`, as documented above, so it can use an empty instance.

`SqliteConnectException` and `SqliteQueryException` on the other hand, each have the `Exception` instance that place them beneath `SomeSqliteException` in the hierarchy.